tools/kani-cov/src/coverage.rs

```diff
@@ -0,0 +1,209 @@
+    pub function: CoverageMetric,
+    pub line: CoverageMetric,
+    pub region: CoverageMetric,
+}
+
+/// A coverage metric.
+pub struct CoverageMetric {
+    pub covered: u32,
+    pub total: u32,
+}
+
+impl CoverageMetric {
+    pub fn new(covered: u32, total: u32) -> Self {
+        CoverageMetric { covered, total }
+    }
+}
+
+/// Function information obtained through a tree-sitter
+#[derive(Debug)]
+pub struct FunctionInfo {
+    pub name: String,
+    pub start: (usize, usize),
+    pub end: (usize, usize),
+}
+
+/// Extract function information from a file using a tree-sitter
+pub fn function_info_from_file(filepath: &PathBuf) -> Vec<FunctionInfo> {
+    let source_code = fs::read_to_string(filepath).expect("could not read source file");
+    let mut parser = Parser::new();
+    parser.set_language(&tree_sitter_rust::language()).expect("Error loading Rust grammar");
+
+    let tree = parser.parse(&source_code, None).expect("Failed to parse file");
+
+    let mut cursor = tree.walk();
+    let first_child_exists = cursor.goto_first_child();
+
+    if !first_child_exists {
+        return vec![];
+    }
+
+    let mut function_info: Vec<FunctionInfo> = Vec::new();
+
+    if cursor.node().kind() == "function_item" {
+        function_info.push(function_info_from_node(cursor.node(), source_code.as_bytes()))
+    };
+
+    while cursor.goto_next_sibling() {
+        if cursor.node().kind() == "function_item" {
+            function_info.push(function_info_from_node(cursor.node(), source_code.as_bytes()))
+        }
+    }
+
+    function_info
+}
+
+/// Helper function to extract function information using a tree-sitter
+fn function_info_from_node(node: Node, source: &[u8]) -> FunctionInfo {
+    let name = node
+        .child_by_field_name("name")
+        .and_then(|name| name.utf8_text(source).ok())
+        .expect("couldn't get function name")
+        .to_string();
+    let start = (node.start_position().row + 1, node.start_position().column + 1);
+    let end = (node.end_position().row + 1, node.end_position().column + 1);
+    FunctionInfo { name, start, end }
+}
+
+/// Extract the coverage results associated to a function
+pub fn function_coverage_results(
+    info: &FunctionInfo,
+    file: &Path,
+    results: &CombinedCoverageResults,
+) -> Option<(String, Vec<CovResult>)> {
+    // The filenames in "kaniraw" files are not absolute, so we need to match
+    // them with the ones we have in the aggregated results (i.e., the filenames
+    // in the "kanimap" files).
+    let filename = file.to_path_buf().into_os_string().into_string().unwrap();
+    let right_filename = results.data.keys().find(|p| filename.ends_with(*p)).unwrap();
+    // TODO: The filenames in kaniraw files should be absolute, just like in metadata.
+    // Otherwise the key for `results` just fails... <https://github.com/model-checking/kani/issues/3542>
+    let file_results = results.data.get(right_filename).unwrap();
+    let function = info.name.clone();
```

Comment:
Why do you need a clone here? The reference might be fine.

Reply:
True, done.